lib/livebook_web/live/hub_live.ex

```diff
@@ -0,0 +1,323 @@
+
+  @impl true
+  def handle_params(%{"id" => id}, _url, socket) do
+    hub = Settings.hub_by_id!(id)
+    hubs = [hub]
+
+    data =
+      case hub.type do
+        "fly" ->
+          %{
+            "organization" => hub.id,
+            "token" => hub.token,
+            "name" => hub.name,
+            "hex_color" => hub.color
+          }
+      end
+
+    opts = select_hub_options(hubs, data["organization"])
+
+    {:noreply,
+     assign(socket,
+       operation: :edit,
+       selected_hub_service: hub.type,
+       data: data,
+       hubs: hubs,
+       hub_options: opts
+     )}
+  end
+
+  def handle_params(_params, _url, socket) do
+    {:noreply, assign(socket, operation: :new)}
+  end
+
+  @impl true
+  def handle_event("select_hub_service", %{"value" => service}, socket) do
+    {:noreply, assign(socket, selected_hub_service: service)}
+  end
+
+  def handle_event("fetch_hubs", %{"fly" => %{"token" => token}}, socket) do
+    case Hub.fetch_hubs(%Fly{token: token}) do
+      {:ok, hubs} ->
+        data = %{"token" => token, "hex_color" => User.random_hex_color()}
+        opts = select_hub_options(hubs)
+
+        {:noreply, assign(socket, data: data, hubs: hubs, hub_options: opts)}
+
+      {:error, _} ->
+        {:noreply,
+         socket
+         |> assign(hubs: [], hub_options: [], data: %{})
+         |> put_flash(:error, "Invalid Access Token")}
+    end
+  end
+
+  def handle_event("save_hub", %{"fly" => params}, socket) do
+    case Enum.find(socket.assigns.hubs, &(&1.id == params["organization"])) do
+      nil ->
+        {:noreply,
+         socket
+         |> assign(data: params)
+         |> put_flash(:error, "Internal Server Error")}
+
+      selected_hub ->
+        case {socket.assigns.operation, Settings.hub_exists?(selected_hub)} do
+          {:new, false} ->
+            {:noreply, save_fly_hub(socket, params, selected_hub)}
+
+          {:edit, true} ->
+            {:noreply, save_fly_hub(socket, params, selected_hub)}
+
+          _any ->
+            {:noreply,
+             socket
+             |> assign(data: params)
+             |> put_flash(:error, "Hub already exists")}
+        end
+    end
+  end
+
+  def handle_event("update_data", %{"fly" => data}, socket) do
+    opts = select_hub_options(socket.assigns.hubs, data["organization"])
+
+    {:noreply, assign(socket, data: data, hub_options: opts)}
+  end
+
+  def handle_event("randomize_color", _, socket) do
+    data = Map.put(socket.assigns.data, "hex_color", User.random_hex_color())
+    {:noreply, assign(socket, data: data)}
+  end
+
+  defp save_fly_hub(socket, params, selected_hub) do
+    user = socket.assigns.current_user
+    opts = select_hub_options(socket.assigns.hubs, params["organization"])
+
+    Settings.save_hub(%{
+      selected_hub
+      | name: params["name"],
+        color: params["hex_color"],
+        token: params["token"]
+    })
+
+    Phoenix.PubSub.broadcast(Livebook.PubSub, "users:#{user.id}", :update_hub)
```

Comment:
We want other users to see the change too and generally it would make sense to have a separate `"hubs"` topic, and we can wrap the subscription in `Livebook.Hubs.subscribe()`. We then need to subscribe in a mount hook for the sidebar.

Also, let's have the message `:hubs_changed`, so it's more clear that the subscriber is expected to just refetch. We could have more specific events for creation/change/remove, but hubs should change very rarely, so refetching sounds good!